File: 92-reverse-linked-list-ii/reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* head) {
        if(head == NULL) return NULL;
        ListNode* p = NULL;
        ListNode* c = head;
        ListNode* n = head->next;
        while(c->next != NULL) {
            c->next=p;
            p=c;
            c=n;
            n=n->next;
        }
        c->next=p;
        return c;
    }
    
    ListNode* reverseBetween(ListNode* head, int left, int right) {

        if(head == NULL || head->next==NULL || left==right) return head;
        ListNode* a = NULL;
        ListNode* b = head;
        ListNode* l=head;
        ListNode* r = NULL;
        for(int i = 1; i < left; i++) {
            a=l;
            l=l->next;
        }
        for(int i = 1; i <= right; i++) {
            r=b;
            b=b->next;
        }
        if(a != NULL) a->next=NULL;
        r->next=NULL;
        ListNode* h = reverse(l);
        l->next=b;
        if(a!=NULL) {
            a->next=h;
        }
        else {
            return r;
        }
        return head;
    }
};

File: 92-reverse-linked-list-ii/reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution 
{
    public ListNode reverseBetween(ListNode head, int left, int right) 
    {
        if(head==null || head.next==null || left==right) return head;
        ListNode a=null;
        ListNode b=head;
        ListNode c=null;
        ListNode d=head;

        for(int i=1;i<left;i++)
        {
            a=b;
            b=b.next;
        }
        for(int i=1;i<=right;i++)
        {
            c=d;
            d=d.next;
        }
        if(a!=null) a.next=null;
        c.next=null;
        ListNode t=iter(b);
        b.next=d;
        if(a!=null) 
        {
            a.next=t;
            return head;
        }
        else
        {
            return t;
        }      
    }
    public ListNode iter(ListNode temp)
    {
        if(temp==null) return null;
        ListNode p=null;
        ListNode c=temp;

        while(c!=null)
        {
            ListNode n=c.next;
            c.next=p;
            p=c;
            c=n;
        }
        return p;
    }
}